### ISSUE TITLE
Outputs without `y_true`

### ISSUE THREAD

**Issue by EderSantana** [fchollet/keras#663]
Sometimes we need to create an output of a graph that does not need a desired. For example, variational auto-encoders have a gaussian prior that needs to be optimized that are solely function of the hidden states. Sparse auto-encoders have the same property. Also, we may want to create a measurement output to visualize what is going on inside the network (like how sparse is a hidden state, what is the spectral radius of a recurrent layer, etc).

Right now, I'm solving this problem by creating an output and defining an `objective` function where `y_true` is merely a place holder. Nevertheless, I still have to pass a value for `y_true` at `fit` time and take space on the GPU.

My question is, should we write a new type of output that does not need a desired or does Keras already have a way to solve that problem?

**Comment by wuaalb** [comment]
I have been wondering how one could implement the variational auto-encoder in Keras, so I think your issue is of interest to me.
However, from your post I'm not completely sure what problem you're talking about exactly, or what solution you propose. Would you mind giving some more details?

When I was thinking about it, I found the main problem to be that Keras loss functions have the fixed form `loss = f(y_true, y_pred)` while the VAE has a loss function with the form `loss = f(theta, phi, x)`. Here `theta` and `phi` are parameters of probability distributions, which in a graph representation might correspond to several nodes each (ie. outputs of layers rather than model parameters).

I didn't see any obvious way to do this in the current framework..

**Comment by EderSantana** [comment]
> When I was thinking about it, I found the main problem to be that Keras loss functions have the fixed form `loss = f(y_true, y_pred)` while the VAE has a loss function with the form `loss = f(theta, phi, x)`.

The VAE loss can be `loss = f(theta, phi, x)` but the mean and log-sigma can come out of a layer. You can merge them with `concat` and split inside the cost function definition, see how I do this here https://github.com/EderSantana/seya/blob/master/seya/objectives.py#L12-L19. 
Nevertheless, you would still have to pass an `y_true` (I call it `dumb` in my code), for instance. The (conv-)VAE with Keras works. But I had to use this trick. On that same problem, I suggested a `join`  merge_mode, so we wouldn't have to handle clumsy splitting manually.

**Comment by wuaalb** [comment]
Ok, thanks, I see what your graph looks like now and why you want this feature.
If it would be straight forward to implement, I agree it would be a little nicer.

About the 'join' merge mode; would that also work if you used >1 samples for the Monte Carlo estimate of the expected reconstruction error for Gaussian data?

**Comment by EderSantana** [comment]
yeah, your layer can output a list just as big as you need with several samples. You can do that already, you know. Just repeat my trick for each output.

**Comment by wuaalb** [comment]
What about instead of Graph output loss doing your `gaussianKL` as a `Regularizer` (see `ActivityRegularizer`)?

**Comment by EderSantana** [comment]
Well thought!!! I'll try that out! I'll let you know later. Thanks!!!

**Comment by EderSantana** [comment]
@wuaalb the problem can be solved with your suggestion. Check this out #689
